### PR TITLE
Convert to promise-based middleware

### PR DESCRIPTION
**Pull request by smcmurray**
This uses the koa v2 series (currently alpha) and the koa-send v3 series.

### CODE REVIEW

**Comment by smcmurray** [comment]
The travis failures are on node pre-4 instances. Koa 2 uses classes, arrow functions, etc. not supported in those versions of node.

**Comment by jonathanong** [comment]
@smcmurray feel free to remove those versions from the travis file!

**Comment by smcmurray** [comment]
@jonathanong It passes all checks now.

File: index.js [hunk 1]
```diff
@@ -36,21 +36,22 @@ function serve(root, opts) {
   opts.index = opts.index || 'index.html';
 
   if (!opts.defer) {
-    return function *serve(next){
-      if (this.method == 'HEAD' || this.method == 'GET') {
-        if (yield send(this, this.path, opts)) return;
+    return function(ctx, next){
```
Comment (coderhaoxin) [inline index.js]: keep function `serve`?

`return function serve(ctx, next){`

**Comment by smcmurray** [comment]
@coderhaoxin , I put the function name back in for you. Out of curiosity, what value does this provide?

**Comment by tejasmanohar** [comment]
@smcmurray readability

File: index.js [hunk 2]
```diff
@@ -36,21 +36,22 @@ function serve(root, opts) {
   if (opts.index !== false) opts.index = opts.index || 'index.html';
 
   if (!opts.defer) {
-    return function *serve(next){
-      if (this.method == 'HEAD' || this.method == 'GET') {
-        if (yield send(this, this.path, opts)) return;
+    return function serve(ctx, next){
+      if (ctx.method == 'HEAD' || ctx.method == 'GET') {
+        return send(ctx, ctx.path, opts).then(function(res){
+          return res ? null : next(ctx);
```
Comment (coderhaoxin) [inline index.js]: no need pass `ctx` to `next()`

Comment (tejasmanohar) [inline index.js]: yea, used to have to do this but then @jonathanong changed the api iirc

Comment (smcmurray) [inline index.js]: @coderhaoxin , @tejasmanohar When did @jonathanong change this? As far as I can tell, ctx is a required argument in koa 2.

Comment (jonathanong) [inline index.js]: no i didn'pt change that. i was just reading the question wrong - i thought he was asking about `async (ctx, next) => {}`

Comment (tejasmanohar) [inline index.js]: my b

**Comment by coderhaoxin** [comment]
@jonathanong
Should we migrate to `koa@2` util `async&await` is native?

**Comment by dead-horse** [comment]
@coderhaoxin  see https://github.com/koajs/koa/issues/603

**Comment by tejasmanohar** [comment]
should prob add this to `next` branch ?

**Comment by coderhaoxin** [comment]
according to https://github.com/koajs/koa/issues/625, I'd like to close this for now :cry:

**Pull request closed without merging**